### PR TITLE
`azurerm_trusted_signing_account` - new resource support

### LINKED ISSUE #26388: Support for Azure Trusted Signing Account
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

There is a new service in Azure for code signing - Trusted Signing Account
https://learn.microsoft.com/en-us/azure/trusted-signing/concept-trusted-signing-resources-roles

It would be good to have this available on terraform.

### New or Affected Resource(s)/Data Source(s)

azurerm_trusted_signing_account

### Potential Terraform Configuration

_No response_

### References

https://learn.microsoft.com/en-us/azure/trusted-signing/concept-trusted-signing-resources-roles

### PR DESCRIPTION
**Pull request by ziyeqf**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Add new resource `azurerm_trusted_signing_account` per the document: https://learn.microsoft.com/en-us/azure/trusted-signing/concept-trusted-signing-resources-roles
The resource provider is named "Microsoft.Codesigning", however the name on Microsoft document is "Trusted signing", so I named this resource to `azurerm_trusted_signing_account"

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

```
❯❯ tftest codesigning TestAccTrustedSigningAccount
=== RUN   TestAccTrustedSigningAccount_basic
=== PAUSE TestAccTrustedSigningAccount_basic
=== RUN   TestAccTrustedSigningAccount_update
=== PAUSE TestAccTrustedSigningAccount_update
=== RUN   TestAccTrustedSigningAccount_complete
=== PAUSE TestAccTrustedSigningAccount_complete
=== CONT  TestAccTrustedSigningAccount_basic
=== CONT  TestAccTrustedSigningAccount_complete
=== CONT  TestAccTrustedSigningAccount_update
--- PASS: TestAccTrustedSigningAccount_update (412.25s)
--- PASS: TestAccTrustedSigningAccount_complete (413.04s)
--- PASS: TestAccTrustedSigningAccount_basic (413.31s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/codesigning	413.386s
```

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_trusted_signing_account` - new resource support [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #26388

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/codesigning/client/client.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	"fmt"
+
+	codesigning_v2024_09_30_preview "github.com/hashicorp/go-azure-sdk/resource-manager/codesigning/2024-09-30-preview"
+
+	"github.com/hashicorp/go-azure-sdk/sdk/client/resourcemanager"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/common"
+)
+
+type Client struct {
+	V20240930previewClient codesigning_v2024_09_30_preview.Client
```
Comment (magodo) [inline internal/services/codesigning/client/client.go:13]: Since this is the only client, can we just name it as `Client` or something?

File: website/docs/r/trusted_signing_account.html.markdown [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+---
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the Trusted Signing.
+
+---
+
+A `sku` block supports the following:
+
+* `name` - (Required) The sku name of this Trusted Signing Account.
```
Comment (magodo) [inline website/docs/r/trusted_signing_account.html.markdown:50]: As this is the only attribute nested, can we flatten it one level upwards and rename it to `sku_name`?

File: internal/services/codesigning/trusted_signing_account_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		},
+		data.ImportStep(),
+		{
+			Config: r.complete(data),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+	})
```
Comment (magodo) [inline internal/services/codesigning/trusted_signing_account_resource_test.go:66]: Can we update it back to `basic`? Otherwise, if a replace is required, can we document it in the docs?

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+					ele := model.Sku[0]
+					upd.Properties.Sku = pointer.To(codesigningaccounts.AccountSkuPatch{})
+					upd.Properties.Sku.Name = pointer.To(codesigningaccounts.SkuName(ele.Name))
+				}
+			}
+			if meta.ResourceData.HasChange("tags") {
+				upd.Tags = pointer.To(model.Tags)
+			}
+
+			if _, err = client.Update(ctx, *id, upd); err != nil {
```
Comment (magodo) [inline internal/services/codesigning/trusted_signing_account_resource.go:215]: Can we use the get-then-put pattern instead?

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

File: internal/services/codesigning/client/client.go (hunk 1 above)
Comment (ziyeqf) [reply internal/services/codesigning/client/client.go]: sure, updated

File: website/docs/r/trusted_signing_account.html.markdown (hunk 2 above)
Comment (ziyeqf) [reply website/docs/r/trusted_signing_account.html.markdown]: updated

File: internal/services/codesigning/trusted_signing_account_resource_test.go (hunk 3 above)
Comment (ziyeqf) [reply internal/services/codesigning/trusted_signing_account_resource_test.go:73]: updated the test case

File: internal/services/codesigning/trusted_signing_account_resource.go (hunk 4 above)
Comment (ziyeqf) [reply internal/services/codesigning/trusted_signing_account_resource.go]: the pattern might not fit since the `client.Update()` requires a `xxxPatch{}` object, the model received from `client.Get()` is `xxxModel{}`, which can only be used with `client.Create()`. While when calling `client.Create()` towards an existing account, it errored 
```
performing Create: unexpected status 400 (400 Bad Request) with error: BadResourceOperation: The resource 'acctest-f0io7' already exists. A resource with the same name cannot be created. Please select a new resource name.
```

**Comment by magodo** [comment]
LGTM now, thank you!

File: go.mod [hunk 5]
```diff
@@ -17,8 +17,8 @@ require (
 	github.com/google/go-cmp v0.6.0
 	github.com/google/uuid v1.6.0
 	github.com/hashicorp/go-azure-helpers v0.70.1
-	github.com/hashicorp/go-azure-sdk/resource-manager v0.20241009.1142232
-	github.com/hashicorp/go-azure-sdk/sdk v0.20241009.1142232
+	github.com/hashicorp/go-azure-sdk/resource-manager v0.20241017.1093842
+	github.com/hashicorp/go-azure-sdk/sdk v0.20241017.1093842
```
Comment (stephybun) [inline go.mod:21]: Updating the `go-azure-sdk` should be done in a separate PR otherwise we end up causing conflicts like has happened here.

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+func (m TrustedSigningAccountResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, meta sdk.ResourceMetaData) error {
+			client := meta.Client.CodeSigning.Client.CodeSigningAccounts
+
+			var model TruestedSigningAccountModel
+			if err := meta.Decode(&model); err != nil {
+				return err
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:90]: These errors should be wrapped
```suggestion
				return fmt.Errorf("decoding: %+v", err)
```

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+			}
+
+			req := codesigningaccounts.CodeSigningAccount{}
+			req.Name = &model.Name
+			req.Location = model.Location
+			req.Properties = pointer.To(codesigningaccounts.CodeSigningAccountProperties{})
+			req.Properties.Sku = pointer.To(codesigningaccounts.AccountSku{
+				Name: codesigningaccounts.SkuName(model.SkuName),
+			})
+			req.Tags = &model.Tags
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:110]: Please update this to
```suggestion
			req := codesigningaccounts.CodeSigningAccount{
				Name: &model.Name
				Location: location.Normalize(model.Location)
				Tags: &model.Tags
				Properties: &codesigningaccounts.CodeSigningAccountProperties{
					Sku: &codesigningaccounts.AccountSku{
						Name: codesigningaccounts.SkuName(model.SkuName),
					}
				}
			}
```

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+		Func: func(ctx context.Context, meta sdk.ResourceMetaData) error {
+			id, err := codesigningaccounts.ParseCodeSigningAccountID(meta.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			client := meta.Client.CodeSigning.Client.CodeSigningAccounts
+			result, err := client.Get(ctx, *id)
+			if err != nil {
+				return err
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:135]:
```suggestion
				return fmt.Errorf("retrieving %s: %+v", id, err)
```

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+			output.Location = model.Location
+			output.Tags = pointer.From(model.Tags)
+			if ptrProp := model.Properties; ptrProp != nil {
+				itemProp := *ptrProp
+				output.AccountUri = pointer.From(itemProp.AccountUri)
+				if ptrSku := itemProp.Sku; ptrSku != nil {
+					itemSku := *ptrSku
+					output.SkuName = string(itemSku.Name)
+				}
+			}
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:158]: We don't error if the model is nil in the read, yet we're doing it twice here. This is also inconsistent with our contributor guidelines and the rest of the provider. Please refactor this 
```suggestion
			var output TruestedSigningAccountModel
			output.Name = id.CodeSigningAccountName
			output.ResourceGroupName = id.ResourceGroupName
			
			if model := result.Model; model != nil {
				output.Location = location.Normalize(model.Location)
				output.Tags = pointer.From(model.Tags)
				
				if props := model.Properties; props != nil {
					output.AccountUri = pointer.From(props.AccountUri)
					if sku := props.Sku; sku != nil {
						output.SkuName = string(sku.Name)
					}
				}
			}
```

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+		Timeout: 10 * time.Minute,
+		Func: func(ctx context.Context, meta sdk.ResourceMetaData) (err error) {
+			client := meta.Client.CodeSigning.Client.CodeSigningAccounts
+			id, err := codesigningaccounts.ParseCodeSigningAccountID(meta.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+			var model TruestedSigningAccountModel
+			if err = meta.Decode(&model); err != nil {
+				return fmt.Errorf("decoding %s: %+v", id, err)
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:176]:
```suggestion
				return fmt.Errorf("decoding: %+v", err)
```

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+			}
+
+			var upd codesigningaccounts.CodeSigningAccountPatch
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:179]: Minor but `upd` is not a known acronym or shorthand for `update` which isn't a long variable name in itself. Can we please be consistent with other resources in the provider and name these one of `payload`, `request`, `param` or spell this out completely to `update`?

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+			if err != nil {
+				return err
+			}
+
+			meta.Logger.Infof("deleting %s", id)
+			client := meta.Client.CodeSigning.Client.CodeSigningAccounts
+			if _, err = client.Delete(ctx, *id); err != nil {
+				return fmt.Errorf("deleting %s: %v", id, err)
+			}
+			return nil
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:214]: Please consider the ordering of variable declarations within the CRUD functions, there is no good reason to reorder this when 99% of the provider follows the sequence suggested below.
```suggestion
			client := meta.Client.CodeSigning.Client.CodeSigningAccounts
			id, err := codesigningaccounts.ParseCodeSigningAccountID(meta.ResourceData.Id())
			if err != nil {
				return err
			}

			meta.Logger.Infof("deleting %s", id)
			
			if _, err = client.Delete(ctx, *id); err != nil {
				return fmt.Errorf("deleting %s: %v", id, err)
			}
			return nil
```

File: internal/services/codesigning/trusted_signing_account_resource_test.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+type TrustedSigningAccountResource struct{}
+
+func (a TrustedSigningAccountResource) Exists(ctx context.Context, client *clients.Client, state *pluginsdk.InstanceState) (*bool, error) {
+	id, err := codesigningaccounts.ParseCodeSigningAccountID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := client.CodeSigning.Client.CodeSigningAccounts.Get(ctx, *id)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving TrustedSigningAccount %s: %+v", id, err)
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource_test.go:29]:
```suggestion
		return nil, fmt.Errorf("retrieving %s %+v", id, err)
```

File: internal/services/codesigning/trusted_signing_account_resource_test.go [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	return utils.Bool(resp.Model != nil), nil
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource_test.go:31]:
```suggestion
	return pointer.To(resp.Model != nil), nil
```

File: internal/services/codesigning/registration.go [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import (
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+)
+
+type Registration struct{}
+
+var _ sdk.TypedServiceRegistration = Registration{}
+
+func (r Registration) AssociatedGitHubLabel() string {
+	return "service/codesigning"
```
Comment (stephybun) [inline internal/services/codesigning/registration.go:15]: We should probably just call this 
```suggestion
	return "service/trustedsigning"
```

File: website/docs/r/trusted_signing_account.html.markdown [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    name = "Basic"
+  }
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name which should be used for this Trusted Signing. Changing this forces a new Trusted Signing Account to be created.
```
Comment (stephybun) [inline website/docs/r/trusted_signing_account.html.markdown:34]: Please update all instances of `Trusted Signing` to `Trusted Signing Account`
```suggestion
* `name` - (Required) The name which should be used for this Trusted Signing Account. Changing this forces a new Trusted Signing Account to be created.
```

File: website/docs/r/trusted_signing_account.html.markdown [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+* `tags` - (Optional) A mapping of tags which should be assigned to the Trusted Signing.
+
+
+## Attributes Reference
+
+In addition to the Arguments listed above - the following Attributes are exported: 
+
+* `id` - The ID of the Trusted Signing.
+
+* `account_uri` - The URI of the trusted signing account which is used during signing files.
```
Comment (stephybun) [inline website/docs/r/trusted_signing_account.html.markdown:53]: Please be consistent with the casing throughout the documentation
```suggestion
* `account_uri` - The URI of the Trusted Signing Account which is used during signing files.
```

**Review by stephybun — CHANGES_REQUESTED** [review]
@ziyeqf please fix up the comments left in-line

**Comment by ziyeqf** [comment]
I'm not sure why the `.github/labeler-pull-request-triage.yml` was changed beyond this RP by `make generate`, seems not related to this PR, so just reverted it.

and here is the new test result:
```
❯❯ tftest codesigning TestAccTrustedSigningAccount
=== RUN   TestAccTrustedSigningAccount_basic
=== PAUSE TestAccTrustedSigningAccount_basic
=== RUN   TestAccTrustedSigningAccount_update
=== PAUSE TestAccTrustedSigningAccount_update
=== RUN   TestAccTrustedSigningAccount_complete
=== PAUSE TestAccTrustedSigningAccount_complete
=== CONT  TestAccTrustedSigningAccount_basic
=== CONT  TestAccTrustedSigningAccount_complete
=== CONT  TestAccTrustedSigningAccount_update
--- PASS: TestAccTrustedSigningAccount_basic (412.82s)
--- PASS: TestAccTrustedSigningAccount_complete (413.86s)
--- PASS: TestAccTrustedSigningAccount_update (542.41s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/codesigning	542.497s
```

File: internal/services/codesigning/trusted_signing_account_resource.go [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+			if !response.WasNotFound(existing.HttpResponse) {
+				if err != nil {
+					return fmt.Errorf("retrieving %s: %v", id, err)
+				}
+				return meta.ResourceRequiresImport(m.ResourceType(), id)
+			}
+
+			req := codesigningaccounts.CodeSigningAccount{
+				Name:     &model.Name,
+				Location: model.Location,
```
Comment (stephybun) [inline internal/services/codesigning/trusted_signing_account_resource.go:106]: Forgot to normalize this one as well, we normalize when building the payload as well as before setting into state
```suggestion
				Location: location.Normalize(model.Location),
```

**Review by stephybun — APPROVED** [review]
Thanks @ziyeqf LGTM 🍀 
image
